File: apps/nihongo_flutter/lib/presentation/srs/srs_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:go_router/go_router.dart';

import '../../domain/entity/srs_card.dart';
import '../../domain/repository/result.dart';
import '../../providers.dart';
import '../../utils/srs_algorithm.dart';
import '../../utils/tts_service.dart';

// Rating: quality 1=Again 2=Hard 3=Good 4=Easy (SM-2 mapped)
const _ratings = [
  _Rating(quality: 1, label: 'Lại',  emoji: '🔁', color: Color(0xFFEF4444)),
  _Rating(quality: 2, label: 'Khó',  emoji: '💪', color: Color(0xFFF97316)),
  _Rating(quality: 3, label: 'Ổn',   emoji: '✅', color: Color(0xFF22C55E)),
  _Rating(quality: 4, label: 'Dễ',   emoji: '⚡', color: Color(0xFF3B82F6)),
];

class _Rating {
  const _Rating({
    required this.quality,
    required this.label,
    required this.emoji,
    required this.color,
  });
  final int quality;
  final String label;
  final String emoji;
  final Color color;
}

// ── Screen ────────────────────────────────────────────────────────────────────

class SrsScreen extends ConsumerStatefulWidget {
  const SrsScreen({super.key});

  @override
  ConsumerState<SrsScreen> createState() => _SrsScreenState();
}

class _SrsScreenState extends ConsumerState<SrsScreen> {
  bool _showAnswer = false;
  bool _speaking = false;
  bool _submitting = false;
  int _sessionCorrect = 0;
  int _sessionTotal = 0;

  // _done is derived from the stream: cards.isEmpty && _sessionTotal > 0
  // No local _done flag — avoids race between stream rebuild and local state.

  Future<void> _rate(int quality, List<ReviewCard> cards) async {
    if (_submitting || cards.isEmpty) return;
    setState(() => _submitting = true);

    final current = cards.first;
    final updated = SrsAlgorithm.calculateNextReview(current.card, quality);
    final result = await ref.read(vocabRepositoryProvider).updateSrsCard(updated);

    if (!mounted) return;

    if (result is Failure) {
      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(content: Text('Lỗi lưu: ${result.error}')),
      );
    } else {
      if (quality >= 3) _sessionCorrect++;
      _sessionTotal++;
    }

    // Stream từ watchDueReviewCards() tự emit list mới sau khi DB update.
    // Chỉ cần reset UI state; done screen hiện khi stream emit [].
    setState(() {
      _showAnswer = false;
      _submitting = false;
    });
  }

  Future<void> _speak(String text) async {
    setState(() => _speaking = true);
    await TtsService.instance.speak(text);
    if (mounted) setState(() => _speaking = false);
  }

  void _restart() {
    setState(() {
      _showAnswer = false;
      _sessionCorrect = 0;
      _sessionTotal = 0;
    });
  }

  @override
  Widget build(BuildContext context) {
    final cardsAsync = ref.watch(reviewQueueProvider);
    final isOnline = ref.watch(isOnlineProvider);
    final scheme = Theme.of(context).colorScheme;

    return Scaffold(
      appBar: AppBar(
        title: const Text('Ôn tập SRS'),
        actions: [
          isOnline.when(
            data: (online) => online
                ? const SizedBox.shrink()
                : Padding(
                    padding: const EdgeInsets.only(right: 8),
                    child: Chip(
                      avatar: const Icon(Icons.cloud_off, size: 16),
                      label: const Text('Offline'),
                      backgroundColor: scheme.errorContainer,
                    ),
                  ),
            loading: () => const SizedBox.shrink(),
            error: (_, __) => const SizedBox.shrink(),
          ),
        ],
      ),
      body: cardsAsync.when(
        loading: () => const Center(child: CircularProgressIndicator()),
        error: (e, _) => Center(child: Text('Lỗi: $e')),
        data: (cards) {
          if (cards.isEmpty) {
            if (_sessionTotal == 0) {
              return _EmptyView();
            }
            return _DoneView(
              correct: _sessionCorrect,
              total: _sessionTotal,
              onRestart: _restart,
            );
          }

          final card = cards.first;
          final displayText =
              (card.kanji != null && card.kanji!.isNotEmpty) ? card.kanji! : card.kana;

          return Column(
            children: [
              // Progress bar
              _ProgressBar(
                current: _sessionTotal,
                total: _sessionTotal + cards.length,
                remaining: cards.length,
              ),

              // Flashcard
              Expanded(
                child: GestureDetector(
                  onTap: !_showAnswer ? () => setState(() => _showAnswer = true) : null,
                  child: Padding(
                    padding: const EdgeInsets.fromLTRB(16, 12, 16, 0),
                    child: _FlashCard(
                      displayText: displayText,
                      kana: card.kana,
                      meaning: card.meaning,
                      romaji: card.romaji,
                      showAnswer: _showAnswer,
                      speaking: _speaking,
                      onSpeak: () => _speak(card.kana),
                      onPronounce: () => context.push(
                        '/pronunciation',
                        extra: {'kana': card.kana, 'meaning': card.meaning},
                      ),
                    ),
                  ),
                ),
              ),

              // Buttons
              AnimatedSwitcher(
                duration: const Duration(milliseconds: 200),
                child: !_showAnswer
                    ? _ShowAnswerBtn(
                        key: const ValueKey('show'),
                        onTap: () => setState(() => _showAnswer = true),
                      )
                    : _RatingRow(
                        key: const ValueKey('rate'),
                        submitting: _submitting,
                        onRate: (q) => _rate(q, cards),
                      ),
              ),

              Text(
                'Còn ${cards.length} thẻ',
                style: Theme.of(context)
                    .textTheme
                    .bodySmall
                    ?.copyWith(color: scheme.onSurfaceVariant),
              ),
              const SizedBox(height: 12),
            ],
          );
        },
      ),
    );
  }
}

// ── Progress bar ──────────────────────────────────────────────────────────────

class _ProgressBar extends StatelessWidget {
  const _ProgressBar({
    required this.current,
    required this.total,
    required this.remaining,
  });
  final int current, total, remaining;

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    final pct = total > 0 ? current / total : 0.0;

    return Padding(
      padding: const EdgeInsets.fromLTRB(16, 8, 16, 0),
      child: Row(
        children: [
          Expanded(
            child: ClipRRect(
              borderRadius: BorderRadius.circular(99),
              child: LinearProgressIndicator(
                value: pct,
                minHeight: 6,
                backgroundColor: scheme.outlineVariant,
                valueColor: AlwaysStoppedAnimation(scheme.primary),
              ),
            ),
          ),
          const SizedBox(width: 12),
          Text(
            '$current/$total',
            style: Theme.of(context).textTheme.labelSmall?.copyWith(
                  color: scheme.onSurfaceVariant,
                  fontVariations: const [FontVariation('wght', 600)],
                ),
          ),
        ],
      ),
    );
  }
}

// ── Flashcard ─────────────────────────────────────────────────────────────────

class _FlashCard extends StatelessWidget {
  const _FlashCard({
    required this.displayText,
    required this.kana,
    required this.meaning,
    required this.romaji,
    required this.showAnswer,
    required this.speaking,
    required this.onSpeak,
    required this.onPronounce,
  });

  final String displayText, kana, meaning, romaji;
  final bool showAnswer, speaking;
  final VoidCallback onSpeak, onPronounce;

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;

    return Card(
      child: SizedBox(
        width: double.infinity,
        child: Stack(
          children: [
            // Main content
            Center(
              child: Padding(
                padding: const EdgeInsets.fromLTRB(24, 56, 24, 24),
                child: Column(
                  mainAxisSize: MainAxisSize.min,
                  children: [
                    // Japanese word
                    Text(
                      displayText,
                      style: Theme.of(context).textTheme.displayMedium?.copyWith(
                            fontWeight: FontWeight.w700,
                            letterSpacing: -1,
                          ),
                      textAlign: TextAlign.center,
                    ),

                    if (displayText != kana) ...[
                      const SizedBox(height: 6),
                      Text(
                        kana,
                        style: Theme.of(context).textTheme.titleLarge?.copyWith(
                              color: scheme.primary,
                            ),
                      ),
                    ],

                    // Answer
                    if (showAnswer) ...[
                      const SizedBox(height: 20),
                      Divider(
                          color: scheme.outlineVariant,
                          indent: 32,
                          endIndent: 32),
                      const SizedBox(height: 20),
                      Text(
                        meaning,
                        style: Theme.of(context)
                            .textTheme
                            .headlineSmall
                            ?.copyWith(fontWeight: FontWeight.w700),
                        textAlign: TextAlign.center,
                      ),
                      const SizedBox(height: 6),
                      Text(
                        romaji,
                        style: Theme.of(context).textTheme.bodyMedium?.copyWith(
                              color: scheme.onSurfaceVariant,
                            ),
                      ),
                    ] else ...[
                      const SizedBox(height: 28),
                      Text(
                        'Nhấn để xem nghĩa',
                        style: Theme.of(context).textTheme.bodySmall?.copyWith(
                              color: scheme.onSurfaceVariant,
                            ),
                      ),
                    ],
                  ],
                ),
              ),
            ),

            // Audio + pronunciation buttons (top-right)
            Positioned(
              top: 12,
              right: 12,
              child: Row(
                children: [
                  IconButton.outlined(
                    onPressed: onPronounce,
                    icon: const Icon(Icons.mic, size: 18),
                    tooltip: 'Luyện phát âm',
                    style: IconButton.styleFrom(
                      minimumSize: const Size(36, 36),
                      padding: EdgeInsets.zero,
                    ),
                  ),
                  const SizedBox(width: 6),
                  IconButton.filled(
                    onPressed: speaking ? null : onSpeak,
                    icon: speaking
                        ? SizedBox(
                            width: 16,
                            height: 16,
                            child: CircularProgressIndicator(
                              strokeWidth: 2,
                              color: scheme.onPrimary,
                            ),
                          )
                        : const Icon(Icons.volume_up, size: 18),
                    tooltip: 'Nghe phát âm',
                    style: IconButton.styleFrom(
                      minimumSize: const Size(36, 36),
                      padding: EdgeInsets.zero,
                    ),
                  ),
                ],
              ),
            ),
          ],
        ),
      ),
    );
  }
}

// ── Show-answer button ────────────────────────────────────────────────────────

class _ShowAnswerBtn extends StatelessWidget {
  const _ShowAnswerBtn({super.key, required this.onTap});
  final VoidCallback onTap;

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.fromLTRB(16, 12, 16, 8),
      child: SizedBox(
        width: double.infinity,
        child: FilledButton(
          onPressed: onTap,
          style: FilledButton.styleFrom(
            padding: const EdgeInsets.symmetric(vertical: 14),
          ),
          child: const Text('Xem đáp án', style: TextStyle(fontSize: 16)),
        ),
      ),
    );
  }
}

// ── Rating row ────────────────────────────────────────────────────────────────

class _RatingRow extends StatelessWidget {
  const _RatingRow({super.key, required this.submitting, required this.onRate});
  final bool submitting;
  final ValueChanged<int> onRate;

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.fromLTRB(16, 12, 16, 8),
      child: Row(
        children: _ratings.map((r) {
          return Expanded(
            child: Padding(
              padding: const EdgeInsets.symmetric(horizontal: 4),
              child: _RatingBtn(
                rating: r,
                submitting: submitting,
                onTap: () => onRate(r.quality),
              ),
            ),
          );
        }).toList(),
      ),
    );
  }
}

class _RatingBtn extends StatelessWidget {
  const _RatingBtn({
    required this.rating,
    required this.submitting,
    required this.onTap,
  });
  final _Rating rating;
  final bool submitting;
  final VoidCallback onTap;

  @override
  Widget build(BuildContext context) {
    return Material(
      color: rating.color,
      borderRadius: BorderRadius.circular(12),
      child: InkWell(
        onTap: submitting ? null : onTap,
        borderRadius: BorderRadius.circular(12),
        child: Opacity(
          opacity: submitting ? 0.5 : 1.0,
          child: Padding(
            padding: const EdgeInsets.symmetric(vertical: 10),
            child: Column(
              mainAxisSize: MainAxisSize.min,
              children: [
                Text(rating.emoji, style: const TextStyle(fontSize: 18)),
                const SizedBox(height: 2),
                Text(
                  rating.label,
                  style: const TextStyle(
                    color: Colors.white,
                    fontSize: 13,
                    fontWeight: FontWeight.w700,
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}

// ── Done screen ───────────────────────────────────────────────────────────────

class _DoneView extends StatelessWidget {
  const _DoneView({
    required this.correct,
    required this.total,
    required this.onRestart,
  });
  final int correct, total;
  final VoidCallback onRestart;

  @override
  Widget build(BuildContext context) {
    final pct = total > 0 ? (correct / total * 100).round() : 0;
    final emoji = pct >= 80 ? '🎉' : pct >= 50 ? '💪' : '📚';
    final scheme = Theme.of(context).colorScheme;

    return Center(
      child: Padding(
        padding: const EdgeInsets.all(32),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            Text(emoji, style: const TextStyle(fontSize: 56)),
            const SizedBox(height: 16),
            Text(
              'Xong phiên ôn!',
              style: Theme.of(context)
                  .textTheme
                  .headlineMedium
                  ?.copyWith(fontWeight: FontWeight.w700),
            ),
            const SizedBox(height: 8),
            Text(
              '$correct/$total đúng — $pct%',
              style: Theme.of(context).textTheme.titleMedium?.copyWith(
                    color: scheme.onSurfaceVariant,
                  ),
            ),
            const SizedBox(height: 32),
            FilledButton.icon(
              onPressed: onRestart,
              icon: const Icon(Icons.refresh),
              label: const Text('Ôn tiếp'),
              style: FilledButton.styleFrom(
                padding: const EdgeInsets.symmetric(
                    horizontal: 32, vertical: 14),
              ),
            ),
          ],
        ),
      ),
    );
  }
}

// ── Empty (no cards due) ──────────────────────────────────────────────────────

class _EmptyView extends StatelessWidget {
  const _EmptyView();

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;

    return Center(
      child: Padding(
        padding: const EdgeInsets.all(32),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            const Text('🎉', style: TextStyle(fontSize: 56)),
            const SizedBox(height: 16),
            Text(
              'Không có thẻ nào cần ôn hôm nay!',
              style: Theme.of(context)
                  .textTheme
                  .titleLarge
                  ?.copyWith(fontWeight: FontWeight.w700),
              textAlign: TextAlign.center,
            ),
            const SizedBox(height: 8),
            Text(
              'Quay lại vào ngày mai để tiếp tục.',
              style: Theme.of(context)
                  .textTheme
                  .bodyMedium
                  ?.copyWith(color: scheme.onSurfaceVariant),
              textAlign: TextAlign.center,
            ),
          ],
        ),
      ),
    );
  }
}

File: apps/nihongo-android/app/src/main/java/com/edu/nihongo/presentation/srs/SrsScreen.kt
package com.edu.nihongo.presentation.srs

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.derivedStateOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.edu.nihongo.domain.entity.ReviewCard

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun SrsScreen(
    onBack: () -> Unit,
    viewModel: SrsViewModel = hiltViewModel(),
) {
    val uiState by viewModel.uiState.collectAsStateWithLifecycle()

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Ôn tập SRS") },
                navigationIcon = {
                    IconButton(onClick = onBack) {
                        Icon(Icons.AutoMirrored.Filled.ArrowBack, contentDescription = "Back")
                    }
                },
            )
        },
    ) { padding ->
        when {
            uiState.isDone -> DoneScreen(
                uiState = uiState,
                onRestart = viewModel::resetSession,
                modifier = Modifier
                    .fillMaxSize()
                    .padding(padding),
            )

            uiState.cards.isEmpty() -> EmptyScreen(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(padding),
            )

            else -> {
                val progress by remember(uiState.sessionTotal, uiState.cards.size) {
                    derivedStateOf {
                        val total = uiState.sessionTotal + uiState.cards.size
                        if (total > 0) uiState.sessionTotal.toFloat() / total.toFloat() else 0f
                    }
                }

                Column(
                    modifier = Modifier
                        .fillMaxSize()
                        .padding(padding),
                ) {
                    LinearProgressIndicator(
                        progress = { progress },
                        modifier = Modifier.fillMaxWidth(),
                    )
                    SrsCardContent(
                        card = uiState.cards.first(),
                        onReviewed = viewModel::onCardReviewed,
                        modifier = Modifier.weight(1f),
                    )
                }
            }
        }
    }
}

@Composable
private fun DoneScreen(
    uiState: SrsUiState,
    onRestart: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier = modifier.padding(24.dp),
        verticalArrangement = Arrangement.Center,
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Text("🎉", style = MaterialTheme.typography.displayLarge)
        Spacer(Modifier.height(16.dp))
        Text("Hoàn thành!", style = MaterialTheme.typography.headlineMedium)
        Spacer(Modifier.height(8.dp))
        Text(
            "Đúng ${uiState.sessionCorrect}/${uiState.sessionTotal} thẻ",
            style = MaterialTheme.typography.titleMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        Spacer(Modifier.height(24.dp))
        Button(onClick = onRestart) {
            Text("Bắt đầu lại")
        }
    }
}

@Composable
private fun EmptyScreen(modifier: Modifier = Modifier) {
    Column(
        modifier = modifier.padding(24.dp),
        verticalArrangement = Arrangement.Center,
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Text("📚", style = MaterialTheme.typography.displayLarge)
        Spacer(Modifier.height(16.dp))
        Text(
            "Không có thẻ cần ôn hôm nay",
            style = MaterialTheme.typography.titleMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
    }
}

@Composable
private fun SrsCardContent(
    card: ReviewCard,
    onReviewed: (ReviewCard, Int) -> Unit,
    modifier: Modifier = Modifier,
) {
    var showAnswer by remember(card.card.id) { mutableStateOf(false) }

    Column(
        modifier = modifier
            .fillMaxSize()
            .padding(16.dp),
    ) {
        Card(
            modifier = Modifier
                .fillMaxWidth()
                .weight(1f)
                .clickable(enabled = !showAnswer) { showAnswer = true },
            elevation = CardDefaults.cardElevation(4.dp),
        ) {
            Column(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(24.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.Center,
            ) {
                Text(card.kana, style = MaterialTheme.typography.displayMedium)
                card.kanji?.let {
                    Text(it, style = MaterialTheme.typography.headlineSmall)
                }
                Spacer(Modifier.height(16.dp))
                if (!showAnswer) {
                    Text(
                        "Chạm để xem đáp án",
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                } else {
                    Text(card.meaning, style = MaterialTheme.typography.bodyLarge)
                    if (card.romaji.isNotBlank()) {
                        Text(
                            card.romaji,
                            style = MaterialTheme.typography.bodyMedium,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                    }
                }
            }
        }

        Spacer(Modifier.height(16.dp))

        if (showAnswer) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(8.dp),
            ) {
                listOf(
                    Triple("🔁", "Lại", Color(0xFFEF4444) to 1),
                    Triple("😓", "Khó", Color(0xFFF97316) to 2),
                    Triple("👌", "Ổn", Color(0xFF22C55E) to 3),
                    Triple("⭐", "Dễ", Color(0xFF3B82F6) to 4),
                ).forEach { (emoji, label, colorQuality) ->
                    val (color, quality) = colorQuality
                    RatingButton(
                        emoji = emoji,
                        label = label,
                        color = color,
                        onClick = {
                            onReviewed(card, quality)
                            showAnswer = false
                        },
                        modifier = Modifier.weight(1f),
                    )
                }
            }
        }
    }
}

@Composable
private fun RatingButton(
    emoji: String,
    label: String,
    color: Color,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Button(
        onClick = onClick,
        modifier = modifier,
        colors = ButtonDefaults.buttonColors(containerColor = color),
        contentPadding = PaddingValues(vertical = 12.dp, horizontal = 4.dp),
    ) {
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.Center,
        ) {
            Text(emoji, style = MaterialTheme.typography.bodyLarge)
            Text(label, style = MaterialTheme.typography.labelSmall)
        }
    }
}
